### PR TITLE
Shape Properties

### LINKED ISSUE #38: Shape Properties
@grantcoz 

As a user I want my shapes to have properties for their dimensions, the pen and the brush that will be used to make the shape
# Tasks:
 - Have a shape object have properties of values such as length, width, height, etc.
 - Have the pen that is being used have a color, width, pen style, pen cap style, and joint style
 - Have the brush that is being used have a color and a shape

# Tests:
 - Check to see if the properties of the shape are correct
 - Check to see if the properties of the pen are correct
 - Check to see if the properties of the brush are correct

**Priority:** (Highest/High/Medium/Low/Lowest)

### PR DESCRIPTION
**Pull request by dacruz21**
Closes #38

Implements properties for Geometric and Text shapes, defined in Properties.hpp. The base Shape class automatically calls applyProperties() on the QPainter when any shape is drawn. The applyProperties() method is a virtual method defined in GeometricShape and Text, which applies the relevant struct of properties to the QPainter.

Test:
 - Open the project, make sure that the square is colorful and dashed
 - Make your own shapes in mainwindow.cpp if you want to play :)

### CODE REVIEW

File: shapes/Circle.hpp [hunk 1]
```diff
@@ -1,14 +1,16 @@
...
 namespace jbrush
 {
-class Circle final : public Shape
+class Circle final : public GeometricShape
 {
 public:
     /* Constructors */
-    Circle(const QPoint& center, int radius);
-    Circle(int x, int y, int radius);
+    Circle(const QPoint& center, int radius) : Circle(center, radius, DEFAULT_GEOMETRIC_PROPS) {}
```
Comment (ChrisRzech) [inline shapes/Circle.hpp]: DAY VID cmon, why you inlining

File: shapes/Circle.hpp [hunk 2]
```diff
@@ -1,14 +1,16 @@
...
+    Circle(int x, int y, int radius) : Circle(x, y, radius, DEFAULT_GEOMETRIC_PROPS) {}
```
Comment (ChrisRzech) [inline shapes/Circle.hpp]: Inline 👎

File: shapes/GeometricShape.hpp [hunk 3]
```diff
@@ -0,0 +1,16 @@
+#pragma once
+
+#include "Shape.hpp"
+#include "Properties.hpp"
+
+namespace jbrush
+{
+class GeometricShape : public Shape
```
Comment (ChrisRzech) [inline shapes/GeometricShape.hpp]: ``GeoShape`` 👍 
``GeometricShape`` 👎

File: shapes/GeometricShape.hpp [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+{
+protected:
+    GeometricShape(const QPoint& pos, GeometricShapeProperties props) : Shape(pos), props(props) {}
```
Comment (ChrisRzech) [inline shapes/GeometricShape.hpp]: inline 👎

File: shapes/Properties.hpp [hunk 5]
```diff
@@ -0,0 +1,43 @@
+#pragma once
+
+#include <QPen>
+#include <QFont>
+
+namespace jbrush
+{
+struct GeometricShapeProperties
```
Comment (ChrisRzech) [inline shapes/Properties.hpp]: ``GeoShapeProp`` 👍 
``GeometricShapeProperties`` 👎

File: shapes/Properties.hpp [hunk 6]
```diff
@@ -0,0 +1,43 @@
+#pragma once
+
```
Comment (ChrisRzech) [inline shapes/Properties.hpp]:
```suggestion
```

File: shapes/GeometricShape.hpp [hunk 7]
```diff
@@ -0,0 +1,16 @@
+#pragma once
+
```
Comment (ChrisRzech) [inline shapes/GeometricShape.hpp]:
```suggestion
```

File: shapes/Polygon.hpp [hunk 8]
```diff
@@ -1,13 +1,14 @@
...
 
 namespace jbrush
 {
-class Polygon final : public Shape
+class Polygon final : public GeometricShape
 {
 public:
     /* Constructors */
-    Polygon(const QPoint* const points, uint32_t pointCount);
+    Polygon(const QPoint* const points, uint32_t pointCount) : Polygon(points, pointCount, DEFAULT_GEOMETRIC_PROPS) {}
```
Comment (ChrisRzech) [inline shapes/Polygon.hpp]: Inline 👎

File: shapes/Polyline.hpp [hunk 9]
```diff
@@ -1,13 +1,14 @@
...
 
 namespace jbrush
 {
-class Polyline final : public Shape
+class Polyline final : public GeometricShape
 {
 public:
     /* Constructors */
-    Polyline(const QPoint* const points, uint32_t pointCount);
+    Polyline(const QPoint* const points, uint32_t pointCount) : Polyline(points, pointCount, DEFAULT_GEOMETRIC_PROPS) {}
```
Comment (ChrisRzech) [inline shapes/Polyline.hpp]: Inline 👎

File: shapes/Properties.hpp (hunk 6 above)
Comment (ChrisRzech) [inline shapes/Properties.hpp]:
```suggestion
```

File: shapes/Rectangle.hpp [hunk 10]
```diff
@@ -0,0 +1,37 @@
+#pragma once
+#include "GeometricShape.hpp"
+
+namespace jbrush
+{
+class Rectangle final : public GeometricShape
+{
+public:
+    /* Constructors */
+    Rectangle(const QPoint& topL, const QPoint& bottomR) : Rectangle(topL, bottomR, DEFAULT_GEOMETRIC_PROPS) {}
```
Comment (ChrisRzech) [inline shapes/Rectangle.hpp]: INLINES EVERYWHERE REEEEEEEEEEEEE 👎

File: shapes/Circle.hpp (hunk 1 above)
Comment (dacruz21) [reply shapes/Circle.hpp]: it's a default constructor u dip

File: shapes/Circle.hpp (hunk 2 above)
Comment (dacruz21) [reply shapes/Circle.hpp]: it's a default constructor u dip

File: shapes/GeometricShape.hpp (hunk 3 above)
Comment (dacruz21) [reply shapes/GeometricShape.hpp]: wtf

File: shapes/GeometricShape.hpp (hunk 4 above)
Comment (dacruz21) [reply shapes/GeometricShape.hpp]: it's a default constructor u dip

File: shapes/Properties.hpp (hunk 5 above)
Comment (dacruz21) [reply shapes/Properties.hpp]: wtf

File: shapes/Polygon.hpp (hunk 8 above)
Comment (dacruz21) [reply shapes/Polygon.hpp]: it's a default constructor u dip

File: shapes/Polyline.hpp (hunk 9 above)
Comment (dacruz21) [reply shapes/Polyline.hpp]: it's a default constructor u dip

File: shapes/Rectangle.hpp (hunk 10 above)
Comment (dacruz21) [reply shapes/Rectangle.hpp]: it's a default constructor u dip

File: shapes/Shape.hpp [hunk 11]
```diff
@@ -45,6 +45,7 @@ class Shape
     /* Shape Data */
     QPoint position;
 
+    virtual void applyProperties(QPainter&) const = 0;
```
Comment (ChrisRzech) [inline shapes/Shape.hpp]: functions needs to breathe
```suggestion
    virtual void applyProperties(QPainter&) const = 0;
    
```

File: shapes/Shape.cpp [hunk 12]
```diff
@@ -25,6 +25,11 @@ void Shape::setPosition(const QPoint& point)
     position = point;
 }
 
+void Shape::draw(QPainter& qp) const
```
Comment (ChrisRzech) [inline shapes/Shape.cpp]: I would just leave it as pure virtual and call ``applyProperties`` within the derived function's ``draw()``.

Comment (dacruz21) [reply shapes/Shape.cpp]: nah, that makes it hard to forget

Comment (ChrisRzech) [reply shapes/Shape.cpp]: But you call it anyway ``Shape::draw(qp);`` in each shape

File: mainwindow.cpp [hunk 13]
```diff
@@ -1,12 +1,14 @@
...
-#include "shapes/Rect.hpp"
+#include "shapes/Ellipse.hpp"
+#include "shapes/Rectangle.hpp"
+#include "shapes/Properties.hpp"
 
 using jbrush::Canvas;
 
 #include "Canvas.hpp"
-#include "shapes/Rect.hpp"
+#include "shapes/Rectangle.hpp"
```
Comment (ChrisRzech) [inline mainwindow.cpp]: Uh oh, redundant include

File: mainwindow.cpp [hunk 14]
```diff
@@ -1,12 +1,14 @@
 #include "mainwindow.hpp"
 #include "ui_mainwindow.h"
 #include "Canvas.hpp"
-#include "shapes/Rect.hpp"
+#include "shapes/Ellipse.hpp"
```
Comment (ChrisRzech) [inline mainwindow.cpp]: Can we just put a ``shapes/Shapes.hpp``

File: shapes/GeometricShape.cpp [hunk 15]
```diff
@@ -0,0 +1,16 @@
+#include "GeometricShape.hpp"
+
+namespace jbrush
+{
+void GeometricShape::applyProperties(QPainter& qp) const
+{
+    QPen pen;
+    pen.setColor(this->props.penColor);
```
Comment (ChrisRzech) [inline shapes/GeometricShape.cpp]: ``this->`` that these those

File: shapes/Shape.cpp [hunk 16]
```diff
@@ -25,6 +25,11 @@ void Shape::setPosition(const QPoint& point)
...
+{
+    this->applyProperties(qp);
```
Comment (ChrisRzech) [inline shapes/Shape.cpp]: ``this->`` that these those

File: mainwindow.cpp [hunk 17]
```diff
@@ -1,12 +1,12 @@
...
 #include "ui_mainwindow.h"
 #include "Canvas.hpp"
-#include "shapes/Rect.hpp"
+#include "shapes/Shapes.hpp"
 
 using jbrush::Canvas;
 
 #include "Canvas.hpp"
-#include "shapes/Rect.hpp"
+#include "shapes/Rectangle.hpp"
```
Comment (ChrisRzech) [inline mainwindow.cpp]: Still redundant

File: shapes/Rectangle.hpp [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+    Rectangle(const QPoint& topL, int width, int height) : Rectangle(topL, width, height, DEFAULT_GEOMETRIC_PROPS) {}
+    Rectangle(int x, int y, int width, int height) : Rectangle(x, y, width, height, DEFAULT_GEOMETRIC_PROPS) {}
+    Rectangle(const QPoint& topL, const QPoint& bottomR, GeometricShapeProperties);
```
Comment (ChrisRzech) [inline shapes/Rectangle.hpp]: Let's just make them actual default parameters then if it's going to look like this.
```suggestion
    Rectangle(const QPoint& topL, const QPoint& bottomR, GeometricShapeProperties = DEFAULT_GEOMETRIC_PROPS);
```

File: shapes/Circle.cpp [hunk 19]
```diff
@@ -48,12 +49,6 @@ std::string Circle::toString() const
            std::to_string(position.x()) + ", " +
```
Comment (BenjiBoy926) [inline shapes/Circle.cpp]:
```suggestion
           std::to_string(position.x()) + jconstants::DIMENSION_INTERRUPTOR + ' ' +
```
Try this for added style and modifiability.  Same for all `toString()` methods

File: shapes/FilledShape.cpp [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+        case Qt::BrushStyle::VerPattern:
+            brushStyle = "VerPattern"; break;
+        case Qt::BrushStyle::NoBrush:
+            brushStyle = "NoBrush"; break;
+        default:
+            throw GeneralException("invalid brush style");
+    }
+
+    return geometricProps + "\n" +
+           "BrushColor: " + brushColor + "\n" +
```
Comment (BenjiBoy926) [inline shapes/FilledShape.cpp]:
```suggestion
           "BrushColor" + jconstants::TITLE_INTERRUPTOR + ' ' + brushColor + "\n" +
```
Try this for added style and modifiability.  Same goes for all `toString()` and `stringifyProperties()` methods

**Pull request merged**